Parsimony/OptionSpec.cs

```diff
@@ -0,0 +1,126 @@
+﻿using System;
+
+namespace Parsimony
+{
+    /// <summary>
+    /// A specification for a CLI option.
+    /// </summary>
+    /// <typeparam name="TOptions">The type of the option set.</typeparam>
+    /// <typeparam name="TValue">The type of this option's value.</typeparam>
+    public class OptionSpec<TOptions, TValue>
+        : OptionSpec<TOptions>
+        where TOptions : notnull, new()
+    {
+        private readonly Func<string, TValue> _parseFn;
+
+        private readonly Action<TOptions, TValue> _setFn;
+
+        /// <summary>
+        /// The default value for the option. If <see cref="OptionSpec{TOptions}.Required"/> is
+        /// <c>true</c> this value  will be ignored.
+        /// </summary>
+        public TValue Default { get; private set; }
+
+        /// <summary>
+        /// The <see cref="Type"/> of the option.
+        /// </summary>
+        public override Type OptionType => typeof(TValue);
+
+        // TODO: Create constructors for valid configurations?
+
+        /// <summary>
+        /// Creates a new <see cref="OptionSpec{TOptions,TValue}"/>.
+        /// </summary>
+        /// <param name="shortName">The short name of the option.</param>
+        /// <param name="longName">The long name of the option.</param>
+        /// <param name="required">Indicates whether the option is required.</param>
+        /// <param name="default">
+        /// The default value for the option (if 'required' is false).
+        /// </param>
+        /// <param name="parseFn">The function to use when parsing the option value.</param>
+        /// <param name="setFn">
+        /// The procedure to use to assign the option value to the <typeparamref name="TOptions"/>.
+        /// </param>
+        /// <exception cref="ArgumentNullException">
+        /// <paramref name="parseFn"/> or <paramref name="setFn"/> is <c>null</c>.
+        /// </exception>
+        /// <exception cref="ArgumentException">
+        /// Both <paramref name="shortName"/> and <paramref name="longName"/> are <c>null</c>.
+        /// </exception>
+        public OptionSpec(
+            char? shortName,
+            string? longName,
+            bool required,
```

Comment:
yea so this type will probably either
- become internal once the parser builder is in place
- expose multiple ctors that ensure validity

I totally agree it's a bad interface now, but since I'm not 100% where it's going I just left it as POCO as possible

Reply:
Makes sense